feat: support many unary dynamo converters

zewenli98 opened the pull request:
# Description

Implemented all 22 unary dynamo converters, including `EXP`, `LOG`, `SQRT`, `RECIP`, `ABS`, `SIN`, `COS`, `TAN`, `SINH`, `COSH`, `ASIN`, `ACOS`, `ATAN`, `ASINH`, `ACOSH`, `ATANH`, `CEIL`, `FLOOR`, `NOT`, `SIGN`, `ROUND`, `ISINF`.

Fixes #2199 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

narendasan commented on py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:
```diff
@@ -405,7 +405,7 @@ def aten_ops_to_copy_dtype(
 
 
-@dynamo_tensorrt_converter(torch.ops.aten.clone.default)
+@dynamo_tensorrt_converter(torch.ops.aten.clone.default)  # type: ignore[misc]
```
Before this is committed check to see if theres a way to fix this mypy error @gs-olive can you look at the decorator as well?

gs-olive replied:
I recently noticed that `mypy` does not accurately show errors for these decorators. I will look more into this.

The decorator is already as strongly-typed as it can be though, as here:
https://github.com/pytorch/TensorRT/blob/585b2b66e50a9900f513b3aa495ae3ed210351d4/py/torch_tensorrt/dynamo/conversion/converter_registry.py#L64-L69

narendasan commented on py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:
```diff
@@ -420,3 +420,523 @@ def aten_ops_clone(
+    ):
+        input_val = cast_trt_tensor(network, input_val, trt.float32, name)
+
+    return impl.unary.unary_op(
```
Expose these as dedicated functions in impl

narendasan reviewed (commented):
@zewenli98 We need a layer of indirection between aten and impl which is why functions like sign had dedicated implementations.

For each op we need:
1. aten registration for the actual converter in aten_converters.py
2. a dedicated impl function 


So for `aten::cos`
We need: 
1. aten_ops_cos (converison.aten_ops_converters)
2. impl.cos (conversion.impl.unary.ops)

zewenli98 commented:
> @zewenli98 We need a layer of indirection between aten and impl which is why functions like sign had dedicated implementations.
> 
> For each op we need:
> 
> 1. aten registration for the actual converter in aten_converters.py
> 2. a dedicated impl function
> 
> So for `aten::cos` We need:
> 
> 1. aten_ops_cos (converison.aten_ops_converters)
> 2. impl.cos (conversion.impl.unary.ops)


I thought there's almost same code for these converters supported by TensorRT, so I created `unary_op` in  `unary/ops.py`, like `elementwise_op` in `elementwise/ops.py`. I was wondering why dedicated implementations are better?

narendasan commented:
Dedicated implementations are better because they use a standardized interface for all ops. So if I want to make a new converter for say LogSumExp, instead of my code looking like this:

```py
layer = impl.unary_op(in, ..., trt.IUnaryOperation.LOG)
layer = impl.reduce_op(layer.output, ..., trt.IReduceOperation.SUM)
return impl.unary_op(layer.output, ..., trt.IUnaryOperation.EXP)
```

Peoples code would look closer to this:

```py
x = impl.log(in,...)
x = impl.sum(x,...)
return impl.exp(x,...)
```

So a converter writer does not need to have in depth knowledge about tensorrt or its APIs, they can write code more similar to pytorch instead.

zewenli98 commented:
> 

Oh I see. It makes sense. Actually at first I implemented dedicated implementations but after seeing `elementwise_op`, I changed to the current code.  I'll change back and also for elementwise.

narendasan commented:
So there can be a core function which is centralized. So `unary_op` is good, but effectively we just need an alias for specific ops so users dont need to do that themselves

zewenli98 commented on py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:
```diff
@@ -420,3 +420,523 @@ def aten_ops_clone(
+    name: str,
+) -> Union[TRTTensor, Sequence[TRTTensor]]:
+    input_val = args[0]
+    if (isinstance(input_val, TRTTensor)) and (
```
I don't think it's unary specific. This depends on the data types TensorRT supported. For unary ops, the supported types are: https://docs.nvidia.com/deeplearning/tensorrt/operators/docs/Unary.html#data-types

narendasan replied:
Lets put the input validation in either impl.unary.exp or impl.unary.unary_op, probably in the op specific part since it seems to vary between ops

zewenli98 replied:
OK, I moved input validation into ops, like inside of `impl.unary.exp`

zewenli98 commented:
OK, I'll change to something like `aten_ops_cos -> impl.cos -> impl.unary.unary_op(..., trt.IUnaryOperation.COS)`

narendasan commented on py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:
```diff
@@ -420,3 +420,501 @@ def aten_ops_clone(
+    name: str,
+) -> Union[TRTTensor, Sequence[TRTTensor]]:
+    input_val = args[0]
+    if (isinstance(input_val, TRTTensor)) and (
```
Lets put the input validation in either impl.unary.exp or impl.unary.unary_op,  probably in the op specific part since it seems to vary between ops

narendasan reviewed (approved):
LGTM @zewenli98 rebase and should be good to merge

zewenli98 commented:
Do we still need `# type: ignore[misc]`?

narendasan commented:
Its fine for now we can resolve it later

zewenli98 commented:
rebased, but I don't have access to merge.
